src/main/java/com/iota/iri/service/ledger/impl/LedgerServiceImpl.java

```diff
@@ -163,54 +163,76 @@ public boolean isBalanceDiffConsistent(Set<Hash> approvedHashes, Map<Hash, Long>
             visitedTransactions.add(solidEntryPointHash);
         });
 
-        final Queue<Hash> nonAnalyzedTransactions = new LinkedList<>(Collections.singleton(startTransaction));
-        Hash transactionPointer;
-        while ((transactionPointer = nonAnalyzedTransactions.poll()) != null) {
-            try {
-                final TransactionViewModel transactionViewModel = TransactionViewModel.fromHash(tangle,
-                        transactionPointer);
-                if (transactionViewModel.getCurrentIndex() == 0 && visitedTransactions.add(transactionPointer)) {
-                    if (transactionViewModel.getType() == TransactionViewModel.PREFILLED_SLOT) {
-                        return null;
-                    }
-                    if (!milestoneService.isTransactionConfirmed(transactionViewModel, milestoneIndex)) {
+        Stack<Hash> stack = new Stack<>();
+        stack.push(startTransaction);
 
-                        final List<TransactionViewModel> bundleTransactions = bundleValidator.validate(tangle,
-                                snapshotProvider.getInitialSnapshot(), transactionViewModel.getHash());
-
-                        if (bundleTransactions.isEmpty()) {
+        while(!stack.empty()){
+            try {
+                Map<Hash, Long> currentState = new HashMap<>();
+                Hash transactionPointer = stack.peek();
+                TransactionViewModel transactionViewModel = TransactionViewModel.fromHash(tangle, transactionPointer);
+                TransactionViewModel trunk = transactionViewModel.getTrunkTransaction(tangle);
+                TransactionViewModel branch = transactionViewModel.getBranchTransaction(tangle);
+
+                boolean approvedTrunk = (trunk.snapshotIndex() > 0) && (trunk.snapshotIndex() != milestoneIndex);
+                boolean approvedBranch = (branch.snapshotIndex() > 0) && (branch.snapshotIndex() != milestoneIndex);
+                if ((trunk.isEmpty()  || visitedTransactions.contains(trunk.getHash()) || approvedTrunk) &&
```

Comment:
One should note that an empty transaction is the same as the genesis transaction (genesis transaction has all 9's trytes, and all 9s Hash). Then it is fine...

If the transaction was empty because the trunk tx is simply missing (not solid) then we have a problem...

I know you went according to the RFC, I just added a comment on it as well...